### ISSUE TITLE
Implement the TUI

### ISSUE THREAD

**Issue by soywod** [soywod/himalaya#24]
https://github.com/crossterm-rs/crossterm

**Comment by matthiasbeyer** [comment]
[Cursive](https://crates.io/crates/cursive)

**Comment by TornaxO7** [comment]
I'm not a rust developer but I saw this: https://github.com/fdehau/tui-rs It's also based on crossterm but it looks like that it has already some "widgets" for the TUI.

**Comment by soywod** [comment]
Yes I saw it also, it is good you mention it. Thanks!

**Comment by TornaxO7** [comment]
Hello @soywod I'm interested in helping you with the TUI.

But I'm pretty new to rust. I've just finished the book yesterday and I played a little bit with `tui-rs` so my "code quality" sucks probably, but I'm ready to improve that. Do you have any suggestions where I could start (if it's fine for you) and which TUI-crate you'd like to use?

**Comment by soywod** [comment]
First of all, thanks for wanting to help!
I think `tui-rs` is a good choice. For the backend I would go with [crossterm](https://github.com/crossterm-rs/crossterm), but you are free to use whatever.
The only thing I would like to keep is the monorepo. I would like to keep all tools related to Himalaya in this repo (vim plugin, emacs pluging, UI, TUI, GUI etc). For now, Himalaya has only one `Cargo.toml`, I think some refactor is necessary to turn it into a monorepo with [workspaces](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).

**Comment by matthiasbeyer** [comment]
> The only thing I would like to keep is the monorepo.

Yes please! That's a really good idea IMO, splitting up things only add unnecessary complexity!

**Comment by matthiasbeyer** [comment]
> I think some refactor is necessary to turn it into a
>monorepo with workspaces

Why? There's no need for that! Just add a subcommand for launching the tui! And, if you want, make the feature a compile time option.

No need for a workspace IMO!

**Comment by TornaxO7** [comment]
I'd say that I'll create a subdirectory will all the TUI stuff. I'll call it probably `tui`. 

>  Just add a subcommand for launching the tui! 

I agree to that. Calling `himalaya --tui` for example.

**Comment by matthiasbeyer** [comment]
> Calling `himalaya --tui` for example.

Or even less to type: `himalaya tui`

**Comment by soywod** [comment]
> Why? There's no need for that!

I was thinking to use directly Rust functions instead of using the JSON API. So I think it makes sense to have workspaces to manage those crates, what do you think @matthiasbeyer?

**Comment by matthiasbeyer** [comment]
Why not have the GUI feature in the source of himalaya itself?

**Comment by soywod** [comment]
Indeed, makes sense. Let's go for that!

**Comment by TornaxO7** [comment]
> Why not have the GUI feature in the source of himalaya itself?

Because I'd like to separate it from the "backend" of himalaya, so we have a clear structure in the future as well. By the way, do you mean `TUI` instead of `GUI`?

**Comment by matthiasbeyer** [comment]
Ah, yes to all points (GUI was caused by autocorrect on my phone, sorry)

**Comment by TornaxO7** [comment]
@soywod so it's fine for you, if I (try) to implement it in a extra-directory: `src/tui`?

**Comment by soywod** [comment]
Yes, you can go for it!

**Comment by timsofteng** [comment]
Is it any news about it?

**Comment by TornaxO7** [comment]
Yes! Take a look into the draft and scroll down to the last messages.

# TL;DR
The PR-Draft isn't given up. It's just stopping at the moment because:
1. I'm waiting for my new laptop (I don't want to develop on a USB-Stick anymore... neovim is even lagging in the TTY....)
2. We'll improve the API code of himalaya

Then the TUI implementation can be continued!

**Comment by TornaxO7** [comment]
But I'm glad that more people want a TUI for himalaya ^^

**Comment by tgharib** [comment]
In the mean time, I implemented my own TUI in bash. You need `sed`, `sd` and `fzf`. Add the following to your bashrc:

    alias stripcolors="sed 's/\x1B\[[0-9;]\{1,\}[A-Za-z]//g'"

    # $1=email ID
    # $2=flags
    hread() {
      himalaya $2 read -t html $1 | w3m -T text/html
    }

    # $1=flags (ignore $1 in sd capture group)
    hmenu() {
      hread $(himalaya $1 | stripcolors | fzf | sd '^(\d{1,10})\s│.+' '$1') "$1"
    }

Example usage:

```
$ himalaya

UID  │FLAGS │SUBJECT                                                                                                             │SENDER                                             │DATE
6659 │      │Your Twitch Account - Successful Log-in                                                                             │Twitch                                             │<censored>
6658 │      │Your Twitch Login Verification Code                                                                                 │Twitch                                             │<censored>
[ ... ]

$ hread 6659 # Open the email in w3m for easy viewing

$ himalaya -a otheraccount # Pass additional flags

UID  │FLAGS │SUBJECT                                                                                                             │SENDER                                             │DATE
6659 │      │Your Twitch Account - Successful Log-in                                                                             │Twitch                                             │<censored>
6658 │      │Your Twitch Login Verification Code                                                                                 │Twitch                                             │<censored>
[ ... ]

$ hread 6659 "-a otheraccount" # Pass additional flags
```

If you're too lazy to enter the email ID 6659 into a terminal, use hmenu:

```
$ hmenu # Lists all the emails in a TUI, select the email you want to view with arrow keys and enter

$ hmenu "-a otheraccount" # Pass additional flags
```

**Comment by guilhermeprokisch** [comment]
Here my  try using [skim](https://github.com/lotabout/skim) and `w3m`

`sk --ansi -i -c 'himalaya list -s 40' --preview "himalaya read {1} -t html | w3m -T text/html" --tac`

**Comment by timsofteng** [comment]
> I also made my experimental TUI using [skim](https://github.com/lotabout/skim) and `w3m`
> 
> ```
> sk --ansi -c 'himalaya list -s 200' --preview "himalaya read {1} -t html | w3m -T text/html" --bind 'space:execute(himalaya reply {1}),esc:execute(himalaya delete {1}),ctrl-f:execute(himalaya forward {1})' --reverse --header-lines=2
> ```
> 
> Press `space` to reply Press `esc` to delete Press `ctrl+f` to forward a message
> 
> ![135011130-54c9abf8-736a-41a4-8656-a607e29a812e](https://user-images.githubusercontent.com/12011070/135011641-696c9fdc-f5dd-4368-aca6-980541aafb8f.jpeg)

Cool!
Why not fzf?

**Comment by guilhermeprokisch** [comment]
@timsofteng  definitely. There's no specific reason over fzf. Here is an fzf version.
```
mail(){

  FZF_DEFAULT_COMMAND='himalaya list -s 200' \
  fzf --bind 'space:execute(himalaya reply {1}),esc:execute(himalaya delete {1}),ctrl-f:execute(himalaya forward {1})' \
      --preview 'himalaya read {1} -t html | w3m -T text/html' \
      --header-lines=2\
      --header 'Space: Reply, Esc: Delete, Ctrl-f: Foward, Ctrl-c: Exit'\
      --layout=reverse\
      --ansi

}
```

I think would be great to implement some kind of cache Himalaya ( Now it's a bit slow cause every time we need to do an IMAP request)

**Comment by TornaxO7** [comment]
The cache is already planned.

**Comment by guilhermeprokisch** [comment]
Also, some ways to customize the display columns.

**Comment by TornaxO7** [comment]
# Hint
If anyone has more time and motivation (I don't mean it in a bad way) than me, **then you can create your own fork** and implement the TUI because at the moment I'm pretty much exhausted by my university and I'd like to complete my eduacational project first... and... sadly the code in my fork isn't up to date anymore...

I'm sorry about this and everyone who were believing in me doing this... ._.

Currently I can't give an exact date or time when I can get back to this and I also don't want you guys to wait for me although I haven't done that much in the last days/weeks for himalaya.

**Comment by soywod** [comment]
Don't worry @TornaxO7, this is an open source project were people help whenever they have time. No pressure. I personally have more free time these days, I will try to continue your work. Thank you anyway for your initiative!

**Comment by Ypnose** [comment]
What is the status of TUI ? Many people would be interested to use `himalaya` interactively, without vim.

**Comment by soywod** [comment]
> What is the status of TUI ?

For now, the TUI feature is in pending state. I would like to finish other issues before taking back this one (like backends for example https://github.com/soywod/himalaya/issues/295).

> Many people would be interested to use `himalaya` interactively

To be honest, it is not the *main* aim of himalaya. The main purpose of this tool is to interact in a CLI way with your emails (console, script etc). I see the TUI feature as a *bonus*. I hope you understand.

**Comment by Ypnose** [comment]
I understand, thanks for your answer.

**Comment by callumacrae** [comment]
Expanded on @guilhermeprokisch fzf script:

```bash
mail() {
  SEARCH_QUERY='
    if [[ "{q}" = "'\'''\''" || "{q}" = "$(echo {)q}" ]]; then
      script -t 0 -q /dev/null himalaya list -s 100 --max-width $COLUMNS
    else
      if [[ -f "$TMPDIR/fzhim_qry" ]]; then
        query=$(eval "echo {q}")
      else
        query="OR OR SUBJECT {q} FROM {q} BODY {q}"
      fi
      script -t 0 -q /dev/null himalaya search -s 100 --max-width $COLUMNS $query
    fi
  '
  FZF_DEFAULT_COMMAND=$SEARCH_QUERY fzf \
    --disabled \
    --bind 'enter:change-preview:himalaya read {1} -t html | w3m -T text/html -cols $FZF_PREVIEW_COLUMNS -o display_link_number=1' \
    --bind "enter:+reload:sleep 0.1; $SEARCH_QUERY" \
    --bind 'ctrl-j:change-preview()+down' \
    --bind 'ctrl-k:change-preview()+up' \
    --bind 'down:change-preview()+down' \
    --bind 'up:change-preview()+up' \
    --bind 'esc:change-preview()' \
    --bind 'ctrl-n:execute:himalaya reply {1}' \
    --bind "ctrl-n:+reload:sleep 0.1; $SEARCH_QUERY" \
    --bind 'ctrl-a:execute:himalaya attachments {1}' \
    --bind "ctrl-r:reload:$SEARCH_QUERY" \
    --bind "change:change-preview()+reload:sleep 0.25; $SEARCH_QUERY" \
    --bind 'ctrl-q:change-prompt(Query > )+execute-silent:touch $TMPDIR/fzhim_qry' \
    --bind 'ctrl-f:change-prompt(> )+execute-silent:rm -f $TMPDIR/fzhim_qry' \
    --preview-window "top" \
    --header "enter: read email, ctrl-n: reply, ctrl-a: attachments, ctrl-q: query mode, ctrl-f: normal mode, ctrl-r: reload, ctrl-c: exit" \
    --header-lines 2 \
    --ansi
  
  rm -f $TMPDIR/fzhim_qry
}
```

Changes made:

- Himalaya is tricked into thinking it's running in a TTY, so colours are output (it would be really nice if this were possible using a flag!)
- IMAP querying is used for search, not fzf
  - Can toggle between two modes: the raw query or searching from/subject/body 
- preview closes when navigating around , meaning stuff won't be marked as read unintentionally
- removed a couple commands i don't use

It's pretty opinionated and tailored towards how I use email but hopefully some people find this useful!

I haven't looked much into how unsafe the `eval` usage is there, but it seems basically fine to me (i'm unescaping stuff escaped by fzf for execution), especially as it's only the user using the program who is able to input stuff in there.

**Comment by soywod** [comment]
> - Himalaya is tricked into thinking it's running in a TTY, so colours
> are output (it would be really nice if this were possible using a
> flag!)

Does this resource [1] helps you?

> It's pretty opinionated and tailored towards how I use email but
> hopefully some people find this useful!

Thanks for sharing, feel free to also share in the Tips section of the
wiki!

[1]: https://github.com/soywod/himalaya/wiki/Tips:disable-colors

On 27 Mar 2022, at 16:34 (+0200), soywod/himalaya wrote:
> Expanded on @guilhermeprokisch fzf script:
> 
> ```bash
> mail() {
>  SEARCH_QUERY='
>  if [[ "{q}" = "'\'''\''" || "{q}" = "$(echo {)q}" ]]; then
>  script -t 0 -q /dev/null himalaya list -s 100 --max-width $COLUMNS
>  else
>  if [[ -f "$TMPDIR/fzhim_qry" ]]; then
>  query=$(eval "echo {q}")
>  else
>  query="OR OR SUBJECT {q} FROM {q} BODY {q}"
>  fi
>  script -t 0 -q /dev/null himalaya search -s 100 --max-width $COLUMNS $query
>  fi
>  '
>  FZF_DEFAULT_COMMAND=$SEARCH_QUERY fzf \
>  --disabled \
>  --bind 'enter:change-preview:himalaya read {1} -t html | w3m -T text/html -cols $FZF_PREVIEW_COLUMNS -o display_link_number=1' \
>  --bind "enter:+reload:sleep 0.1; $SEARCH_QUERY" \
>  --bind 'ctrl-j:change-preview()+down' \
>  --bind 'ctrl-k:change-preview()+up' \
>  --bind 'down:change-preview()+down' \
>  --bind 'up:change-preview()+up' \
>  --bind 'esc:change-preview()' \
>  --bind 'ctrl-n:execute:himalaya reply {1}' \
>  --bind "ctrl-n:+reload:sleep 0.1; $SEARCH_QUERY" \
>  --bind 'ctrl-a:execute:himalaya attachments {1}' \
>  --bind "ctrl-r:reload:$SEARCH_QUERY" \
>  --bind "change:change-preview()+reload:sleep 0.25; $SEARCH_QUERY" \
>  --bind 'ctrl-q:change-prompt(Query > )+execute-silent:touch $TMPDIR/fzhim_qry' \
>  --bind 'ctrl-f:change-prompt(> )+execute-silent:rm -f $TMPDIR/fzhim_qry' \
>  --preview-window "top" \
>  --header "enter: read email, ctrl-n: reply, ctrl-a: attachments, ctrl-q: query mode, ctrl-f: normal mode, ctrl-r: reload, ctrl-c: exit" \
>  --header-lines 2 \
>  --ansi
> 
> rm -f $TMPDIR/fzhim_qry
> }
> ```
> 
> Changes made:
> 
> - Himalaya is tricked into thinking it's running in a TTY, so colours are output (it would be really nice if this were possible using a flag!)
> - IMAP querying is used for search, not fzf
>  - Can toggle between two modes: the raw query or searching from/subject/body 
> - preview closes when navigating around , meaning stuff won't be marked as read unintentionally
> - removed a couple commands i don't use
> 
> It's pretty opinionated and tailored towards how I use email but hopefully some people find this useful!
> 
> I haven't looked much into how unsafe the `eval` usage is there, but it seems basically fine to me (i'm unescaping stuff escaped by fzf for execution), especially as it's only the user using the program who is able to input stuff in there.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/soywod/himalaya/issues/24#issuecomment-1079944050
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

-- 
Cordialement
Clément DOUIN
Développeur Web Full-Stack
https://soywod.me

**Comment by callumacrae** [comment]
> Does this resource [1] helps you?

Sorry to clarify, it's the other way around - I want colours, himalaya didn't want to give them to me

It's possible to override by setting `$TERM` [in termcolor](https://github.com/BurntSushi/termcolor/blob/dc7e7b93830579716fc180925a17551dee5cfddc/src/lib.rs#L233-L251) but himalaya is [slightly stricter](https://github.com/soywod/himalaya/blob/fe8f9eab7c4f1e7c470ac8cea48594d38580f8dc/cli/src/output/printer_service.rs#L66) so using `script` is required, as far as I can tell

> Thanks for sharing, feel free to also share in the Tips section of the wiki!

Going to add mailboxes support and then will do :)

**Comment by soywod** [comment]
> It's possible to override by setting `$TERM` in termcolor but himalaya
> is slightly stricter so using `script` is required, as far as I can
> tell

Interesting, maybe an argument could be introduced to force colors or
not. If this argument is missing then we can fallback to the auto
way. Feel free to open an issue if you find this relevant!

> Going to add mailboxes support and then will do :)

Great, thank you :)

On 28 Mar 2022, at 10:43 (+0200), soywod/himalaya wrote:
>> Does this resource [1] helps you?
> 
> Sorry to clarify, it's the other way around - I want colours, himalaya didn't want to give them to me
> 
> It's possible to override by setting `$TERM` [in termcolor](https://github.com/BurntSushi/termcolor/blob/dc7e7b93830579716fc180925a17551dee5cfddc/src/lib.rs#L233-L251) but himalaya is [slightly stricter](https://github.com/soywod/himalaya/blob/fe8f9eab7c4f1e7c470ac8cea48594d38580f8dc/cli/src/output/printer_service.rs#L66) so using `script` is required, as far as I can tell
> 
>> Thanks for sharing, feel free to also share in the Tips section of the wiki!
> 
> Going to add mailboxes support and then will do :)
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/soywod/himalaya/issues/24#issuecomment-1080363825
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

-- 
Cordialement
Clément DOUIN
Développeur Web Full-Stack
https://soywod.me

**Comment by callumacrae** [comment]
> Feel free to open an issue if you find this relevant!

Need to learn a bit of Rust first (I've written C++ before, shouldn't be that hard right? 😅) and then hoping to be able to contribute some stuff myself :)

**Comment by piotryordanov** [comment]
@callumacrae Thx for the updated script.
Any idea how to make it work on fish? Or maybe how to run it from inside fish shell?

Thx!

**Comment by callumacrae** [comment]
> @callumacrae Thx for the updated script. Any idea how to make it work on fish? Or maybe how to run it from inside fish shell?
> 
> Thx!

dunno, from a quick google it looks like fish scripting is completely different to zsh / bash and doesn't even try to be interoperable - but i've never used fish before so can't really help much

i'd suggest saving the script in a file with a bash or zsh hashbang and calling it as a script, not a function - although you'll also need to modify the $SHELL variable when you call fzf or it'll try to run the scripts in strings in fish too, and I doubt that works

also, you can find the latest version of the script (with multi-account support and a couple other improvements) here: https://github.com/callumacrae/dotfiles/blob/master/zsh/mail.zsh

**Comment by piotryordanov** [comment]
@callumacrae yeah fish is wildly different. I had actually tried giving it a go before pinging you. But I know it's just a pain 😂

So, I figured that i can create the zsh file, then in fish do `zsh mail.zsh`, making sure I run the mail function at the end.
While, it does run, I am getting this error. 
Is it related to the variables you are using and that i would have to update, or something else?
![image](https://user-images.githubusercontent.com/48773775/162546450-b0d7887f-b6de-4b63-9b50-ee3a6bbc64c3.png)

**Comment by callumacrae** [comment]
@piotryordanov

I've made some changes to the script:

- Works in both bash and zsh
- Sets `SHELL` to `/bin/bash` so that the subcommands always run in bash
- Documented how to set it up in bash/zsh/fish

It wasn't working for you because fzf executes things in the shell specified by `$SHELL`, which for you was fish - passing bash as SHELL fixes the issue.

https://github.com/callumacrae/dotfiles/blob/master/zsh/mail.zsh

(i think future discussion should probably be on my dotfiles repository, not here 😅)

**Comment by soywod** [comment]
> Need to learn a bit of Rust first (I've written C++ before, shouldn't
> be that hard right? 😅)

The ownership system is really special and needs some time to be
understood, even when coming from C++. But for sure, it may help a lot!

> and then hoping to be able to contribute some stuff myself :)

Awesome :)

On 28 Mar 2022, at 11:28 (+0200), soywod/himalaya wrote:
>> Feel free to open an issue if you find this relevant!
> 
> Need to learn a bit of Rust first (I've written C++ before, shouldn't be that hard right? 😅) and then hoping to be able to contribute some stuff myself :)
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/soywod/himalaya/issues/24#issuecomment-1080408799
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

-- 
Cordialement
Clément DOUIN
Développeur Web Full-Stack
https://soywod.me

**Comment by soywod** [comment]
I gather all Himalaya issues to the same [bug tracker](https://todo.sr.ht/~soywod/himalaya), so I transfer the issue [here](https://todo.sr.ht/~soywod/himalaya/23).
